lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -348,12 +348,21 @@ func NewNodeJoinWait(cfg *NodeJoinWaitConfig) (*NodeJoinWait, error) {
 //
 // The Electron app calls this method soon after starting the agent process.
 func (n *NodeJoinWait) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, cluster *clusters.Cluster) (clusters.Server, error) {
-	nodeName, err := n.getNodeNameFromHostUUIDFile(ctx, cluster)
+	// The Electron app aborts the request which calls NodeJoinWait after a timeout, but let's use a
+	// timeout internally as well. Both operations in this struct theoretically block forever if an
+	// error happens elsewhere and gRPC doesn't set a timeout by default.
+	//
+	// If there's a bug in the Electron app and it forgets to abort the request, at least the request
+	// will not hang forever.
```

Comment:
```suggestion
	// The Electron app aborts this request after a timeout, but we set a timeout here as well for
	// to ensure that the request can't hang forever if there's a bug in the request abort logic.
```

Two things to consider:

1. Perhaps move this `timeoutCtx` up into [the endpoint itself](https://github.com/gravitational/teleport/blob/e384604811d9c2f73231efd0a16133d5dbcb993d/lib/teleterm/apiserver/handler/handler_connectmycomputer.go#L59-L73). That way it's clear that the whole endpoint just times out after a minute.
2. I'm not sure how big of a change this would be, but since we're adding timeout logic here anyways, there may be a case for just removing it from the frontend code and using this logic exclusively instead, or having the frontend pass the timeout as a request parameter.

Reply:
> 1. Perhaps move this `timeoutCtx` up into [the endpoint itself](https://github.com/gravitational/teleport/blob/e384604811d9c2f73231efd0a16133d5dbcb993d/lib/teleterm/apiserver/handler/handler_connectmycomputer.go#L59-L73). That way it's clear that the whole endpoint just times out after a minute.

Yeah, I wasn't sure if it'd be best to add a timeout tightly around the operation that actually blocks vs for the whole endpoint.

Since those individual services technically could serve as building blocks for other requests, I guess it's best to add the timeout on the endpoint level.

> 2\. I'm not sure how big of a change this would be, but since we're adding timeout logic here anyways, there may be a case for just removing it from the frontend code and using this logic exclusively instead, or having the frontend pass the timeout as a request parameter.

Now that I think about it, under normal circumstances this would be the way to go. However, the UI needs to be able to distinguish between a timeout and other errors. The JS gRPC client encodes that information as custom properties on the error object. gRPC errors in our app pass through [the Electron context bridge](https://www.electronjs.org/docs/latest/api/context-bridge) which strips all custom properties from Error objects, meaning we'd only have the error message to distinguish between a timeout or not.

So until we revamp how those errors are passed, I think I'm going to keep current implementation as is with regards to the timeout.

I'm still not sure if an endpoint should trouble itself with the deadline, perhaps it should rather return `BadParameter` if no deadline was set on the gRPC request? I'd rather use built-in mechanisms to handle stuff like this vs building my own on top of gRPC.